### DISCUSSION TITLE
Can the tinyUSB examples be adapted for the Arduino IDE? Can they be added to the examples in this repository?

### DISCUSSION THREAD

**Discussion by rtek1000** [adafruit/Adafruit_TinyUSB_Arduino#308] [Q&A]
Hi,

I'm interested in running USB Host HUB on the RP2040 board,

------

To operate with Keyboard and Mouse, and transfer data to STM32 via serial port or I2C, I'll see which one to use later.

I have a [project](https://github.com/rtek1000/Datalogger_2039) with STM32F407, which uses two USB ports in Host mode, one for MSC and one for HID.

The problem is that the USB Host core available for the STM32CubeIDE IDE does not have HUB support, which should be the basics.

And to make things worse, I was informed that there is a problem that can cause the USB port to fail, so I'm looking for a retrofit, as the custom board has already been made, and the MSC part seems to work fine, only the HID part I cannot use a keyboard with a touchpad.

> [The weird thing is now that the enumeration of Low-Speed devices is only working about 50% of the time. Once the Low-Speed device is enumerated I can recieve as many interrupt inputs as I want without any issues. But the other 50% that I connect the Low-Speed device to my HUB the USB core state machine gets Stuck in the enumeration process. ](https://community.st.com/s/question/0D50X0000BHydBQSQZ/stm32f429-discovery-as-usb-host-problems-enumerating-lowspeed-device-over-fullspeed-hub)

I also found a library for STM32 and [USB Host with HUB support, in Keil Studio](https://community.st.com/s/feed/0D53W00002IcpiASAR), but it seems that custom boards are not supported in this IDE, only supports development ready boards, which makes the library useless.

------

According to this list, USB Host is supported by tinyUSB on RP2040 board:
https://github.com/hathach/tinyusb/blob/master/docs/reference/supported.rst

I found the examples, for USB Host HID, which must even operate with the HUB, as the support is for several types of devices:
https://github.com/raspberrypi/pico-examples/tree/master/usb/host/host_cdc_msc_hid

My question is about compiling or adapting this example to the Arduino IDE, as I already have the support that the SDK requests installed in the Arduino IDE:

> sudo apt install cmake gcc-arm-none-eabi libnewlib-arm-none-eabi libstdc++-arm-none-eabi-newlib

**Comment by hathach** [comment]
already supported with included examples
